File: .cursor/hooks/playbook.py
#!/usr/bin/env python3
"""Playbook harness hooks.

Commands (stdout is JSON for Cursor; diagnostics go to stderr):

  gate      preToolUse — deny product UI/route/feature writes until shape is done
  session   sessionStart — inject a short pipeline digest as additional_context
  selftest  run fixtures; exit 1 on failure

Repo root is two levels above this file. Cloud and local checkouts both work.
"""

from __future__ import annotations

import json
import sys
from datetime import date
from pathlib import Path
from typing import Any

try:
    import yaml
except ImportError:  # pragma: no cover
    yaml = None  # type: ignore[assignment]

ROOT = Path(__file__).resolve().parents[2]
STATE_PATH = ROOT / "docs" / "product" / "state.yaml"
GATED_FRAGMENTS = ("/src/app/", "/src/features/")
PLAYBOOK_SKILLS = (
    "next",
    "status",
    "artifacts",
    "journeys",
    "design-system",
    "shape",
    "salvage",
    "setup",
    "customize",
    "ontology",
    "prototype",
    "field-kit",
    "linear-sync",
)


def emit(payload: dict[str, Any]) -> None:
    sys.stdout.write(json.dumps(payload, ensure_ascii=False))
    sys.stdout.write("\n")


def read_stdin() -> str:
    return sys.stdin.read() if not sys.stdin.isatty() else ""


# ── minimal YAML subset, used when pyyaml is unavailable ─────────────────────
# The state file uses a small, documented subset: nested block maps, inline flow
# maps, and block sequences of maps. A gate that guards the build must not
# depend on a third-party library being installed in every environment.


def _scalar(tok: str) -> Any:
    tok = tok.strip()
    if not tok:
        return None
    if tok[0] in "\"'" and tok[-1] == tok[0] and len(tok) > 1:
        return tok[1:-1]
    low = tok.lower()
    if low in ("true", "yes"):
        return True
    if low in ("false", "no"):
        return False
    if low in ("null", "~"):
        return None
    try:
        return int(tok)
    except ValueError:
        return tok


def _split_top(text: str, sep: str = ",") -> list[str]:
    parts, depth, buf = [], 0, ""
    for ch in text:
        if ch in "[{":
            depth += 1
        elif ch in "]}":
            depth -= 1
        if ch == sep and depth == 0:
            parts.append(buf)
            buf = ""
        else:
            buf += ch
    if buf.strip():
        parts.append(buf)
    return parts


def _flow(tok: str) -> Any:
    tok = tok.strip()
    if tok.startswith("{") and tok.endswith("}"):
        out: dict[str, Any] = {}
        for part in _split_top(tok[1:-1]):
            if ":" in part:
                k, _, v = part.partition(":")
                out[k.strip()] = _flow(v)
        return out
    if tok.startswith("[") and tok.endswith("]"):
        return [_flow(p) for p in _split_top(tok[1:-1])]
    return _scalar(tok)


def _strip_comment(line: str) -> str:
    out, quote = "", ""
    for ch in line:
        if quote:
            if ch == quote:
                quote = ""
        elif ch in "\"'":
            quote = ch
        elif ch == "#":
            break
        out += ch
    return out.rstrip()


def _parse_block(lines: list[tuple[int, str]], idx: int, indent: int) -> tuple[Any, int]:
    if idx < len(lines) and lines[idx][1].startswith("- "):
        items: list[Any] = []
        while idx < len(lines) and lines[idx][0] == indent and lines[idx][1].startswith("- "):
            rest = lines[idx][1][2:].strip()
            idx += 1
            if rest.startswith("{") or not (":" in rest and not rest.startswith("{")):
                items.append(_flow(rest))
                continue
            item: dict[str, Any] = {}
            k, _, v = rest.partition(":")
            if v.strip():
                item[k.strip()] = _flow(v)
            else:
                child, idx = _parse_block(lines, idx, lines[idx][0] if idx < len(lines) else indent + 2)
                item[k.strip()] = child
            while idx < len(lines) and lines[idx][0] > indent:
                k2, _, v2 = lines[idx][1].partition(":")
                if v2.strip():
                    item[k2.strip()] = _flow(v2)
                    idx += 1
                else:
                    inner_indent = lines[idx + 1][0] if idx + 1 < len(lines) else lines[idx][0] + 2
                    key2 = k2.strip()
                    idx += 1
                    child, idx = _parse_block(lines, idx, inner_indent)
                    item[key2] = child
            items.append(item)
        return items, idx

    mapping: dict[str, Any] = {}
    while idx < len(lines) and lines[idx][0] == indent:
        line = lines[idx][1]
        if ":" not in line:
            idx += 1
            continue
        key, _, val = line.partition(":")
        key = key.strip()
        if val.strip():
            mapping[key] = _flow(val)
            idx += 1
        else:
            idx += 1
            if idx < len(lines) and lines[idx][0] > indent:
                child, idx = _parse_block(lines, idx, lines[idx][0])
            elif idx < len(lines) and lines[idx][1].startswith("- ") and lines[idx][0] == indent:
                child, idx = _parse_block(lines, idx, indent)
            else:
                child = None
            mapping[key] = child
    return mapping, idx


def mini_yaml_load(text: str) -> Any:
    lines: list[tuple[int, str]] = []
    for raw in text.splitlines():
        stripped = _strip_comment(raw)
        if not stripped.strip():
            continue
        lines.append((len(stripped) - len(stripped.lstrip()), stripped.strip()))
    if not lines:
        return {}
    value, _ = _parse_block(lines, 0, lines[0][0])
    return value


def load_yaml_text(text: str) -> Any:
    """Parse YAML with pyyaml when present, the bundled subset parser otherwise."""
    if yaml is None:
        return mini_yaml_load(text)
    return yaml.safe_load(text)


def load_state(text: str | None = None) -> dict[str, Any] | None:
    """Return parsed state, empty dict on missing file, or None on parse failure."""
    if text is None:
        if not STATE_PATH.is_file():
            return {}
        text = STATE_PATH.read_text(encoding="utf-8")
    try:
        if yaml is None:
            data = mini_yaml_load(text) or {}
        else:
            data = yaml.safe_load(text) or {}
    except Exception:
        return None
    if not isinstance(data, dict):
        return None
    return data


def phase_entries(state: dict[str, Any]) -> list[dict[str, Any]]:
    raw = state.get("phases") or {}
    if not isinstance(raw, dict):
        return []
    out: list[dict[str, Any]] = []
    for key, value in raw.items():
        if isinstance(value, dict):
            entry = dict(value)
            entry["_key"] = key
            out.append(entry)
    return out


def named_phase(state: dict[str, Any], name: str) -> dict[str, Any] | None:
    for entry in phase_entries(state):
        if str(entry.get("name") or "") == name:
            return entry
    return None


def shape_done(state: dict[str, Any]) -> bool:
    entry = named_phase(state, "shape")
    return bool(entry) and str(entry.get("status") or "") == "done"


def ui_writes_override(state: dict[str, Any]) -> str | None:
    raw = state.get("ui_writes")
    if raw is None:
        return None
    value = str(raw).strip().lower()
    if value in ("allow", "deny"):
        return value
    return None


def ui_writes_allowed(state: dict[str, Any] | None) -> tuple[bool, str]:
    """Whether product UI writes are allowed. Missing state file → allow (boilerplate)."""
    if state is None:
        return False, "docs/product/state.yaml is unreadable; refusing product UI writes"
    if state == {}:
        return True, "no product pipeline; boilerplate edits allowed"
    override = ui_writes_override(state)
    if override == "allow":
        return True, "ui_writes: allow"
    if override == "deny":
        return False, "ui_writes: deny"
    if shape_done(state):
        return True, "shape is done; design doc is approved"
    return False, "design doc is not approved (shape is not done)"


def collect_paths(tool_input: Any) -> list[str]:
    paths: list[str] = []

    def walk(node: Any) -> None:
        if isinstance(node, str):
            return
        if isinstance(node, dict):
            for key, value in node.items():
                if key in ("path", "file_path", "filePath", "target_notebook") and isinstance(
                    value, str
                ):
                    paths.append(value)
                else:
                    walk(value)
        elif isinstance(node, list):
            for item in node:
                walk(item)

    walk(tool_input)
    return paths


def is_gated_path(path: str) -> bool:
    normalized = path.replace("\\", "/")
    if not normalized.startswith("/"):
        normalized = "/" + normalized
    return any(fragment in normalized for fragment in GATED_FRAGMENTS)


def gate_from_payload(payload: dict[str, Any], state: dict[str, Any] | None) -> dict[str, Any]:
    tool_input = payload.get("tool_input") or payload.get("input") or {}
    paths = collect_paths(tool_input)
    gated = [p for p in paths if is_gated_path(p)]
    if not gated:
        return {"permission": "allow"}
    allowed, reason = ui_writes_allowed(state)
    if allowed:
        return {"permission": "allow"}
    return {
        "permission": "deny",
        "user_message": (
            "Product UI/routes/features are blocked until the design doc is approved. "
            f"{reason}."
        ),
        "agent_message": (
            "HOOK-DENY: writes under apps/*/src/app or apps/*/src/features are blocked "
            f"until shape is done (or ui_writes: allow). {reason}. "
            "Do not retry the same edit or bypass it with the shell. "
            "Work in docs/product, docs/plans, docs/journeys, or docs/research instead."
        ),
    }


def days_open(raised: str) -> str | None:
    try:
        start = date.fromisoformat(str(raised)[:10])
    except ValueError:
        return None
    delta = (date.today() - start).days
    if delta <= 0:
        return "today"
    if delta == 1:
        return "1 day"
    return f"{delta} days"


def compact_where(state: dict[str, Any]) -> str:
    entries = phase_entries(state)
    done = [str(e.get("name")) for e in entries if str(e.get("status") or "") == "done"]
    current = None
    for entry in entries:
        if str(entry.get("status") or "") in ("in-progress", "blocked"):
            current = str(entry.get("name"))
            break
    if current is None:
        current_key = str(state.get("phase") or "")
        current = current_key
        for entry in entries:
            if str(entry.get("_key")) == current_key:
                current = str(entry.get("name") or current_key)
                break
    next_name = None
    seen_current = False
    for entry in entries:
        name = str(entry.get("name") or "")
        if name == current:
            seen_current = True
            continue
        if seen_current and str(entry.get("status") or "") == "pending":
            next_name = name
            break
    parts = []
    if done:
        parts.append("done: " + ", ".join(done))
    if current:
        parts.append("now: " + str(current))
    if next_name:
        parts.append("next: " + next_name)
    return " · ".join(parts) if parts else "in progress"


def status_digest(state: dict[str, Any] | None) -> str:
    if state is None:
        return (
            "Pipeline state file exists but could not be parsed. "
            "Do not edit product UI until it is readable."
        )
    if state == {}:
        return (
            "There is no product being shaped here. App source may be edited (boilerplate). "
            "Offer to start a product if they want one."
        )
    lines: list[str] = []
    held = state.get("held") or []
    open_items = [
        h
        for h in held
        if isinstance(h, dict) and str(h.get("status") or "") == "open"
    ]
    if open_items:
        lines.append("Waiting on you")
        for item in open_items:
            what = str(item.get("what") or "a decision or facts").rstrip(".")
            age = days_open(str(item.get("raised") or ""))
            suffix = f" (open {age})" if age else ""
            lines.append(f"- {what}{suffix}")
    where = compact_where(state)
    if where:
        lines.append("Where we are")
        lines.append(where)
    deferred = [
        h
        for h in held
        if isinstance(h, dict) and str(h.get("status") or "") == "deferred"
    ]
    if deferred:
        lines.append("Deferred")
        for item in deferred:
            what = str(item.get("what") or "a later choice").rstrip(".")
            until = item.get("until")
            when = f" — again {until}" if until else ""
            lines.append(f"- {what}{when}")
    allowed, reason = ui_writes_allowed(state)
    lines.append("Do not" if not allowed else "App source")
    if allowed:
        lines.append(f"Writes under app routes/features are allowed ({reason}).")
    else:
        lines.append(
            "Do not edit apps/*/src/app or apps/*/src/features. "
            f"{reason}. A hook denies those writes."
        )
    lines.append("Harness")
    lines.append(
        "If next/status/shape are missing from your skill list, read "
        ".agents/skills/<name>/SKILL.md (also .cursor/skills/). "
        "Say that the catalog omitted them."
    )
    return "\n".join(lines)


def cmd_gate() -> int:
    raw = read_stdin()
    try:
        payload = json.loads(raw) if raw.strip() else {}
    except json.JSONDecodeError:
        payload = {}
    if not isinstance(payload, dict):
        payload = {}
    state = load_state()
    emit(gate_from_payload(payload, state))
    return 0


def cmd_session() -> int:
    read_stdin()
    emit({"additional_context": status_digest(load_state())})
    return 0


def _payload(path: str) -> dict[str, Any]:
    return {"tool_name": "StrReplace", "tool_input": {"path": path}}


def cmd_selftest() -> int:
    failures: list[str] = []

    def check(name: str, got: Any, expected: Any) -> None:
        if got != expected:
            failures.append(f"{name}: got {got!r} expected {expected!r}")

    pending = load_yaml_text(
        """
phase: 3
phases:
  3: { name: shape, status: pending }
"""
    )
    done = load_yaml_text(
        """
phase: 4
phases:
  3: { name: shape, status: done }
  4: { name: journeys, status: in-progress }
"""
    )
    deny_override = load_yaml_text(
        """
ui_writes: deny
phases:
  3: { name: shape, status: done }
"""
    )
    allow_override = load_yaml_text(
        """
ui_writes: allow
phases:
  3: { name: shape, status: pending }
"""
    )

    app = "/workspace/apps/web/src/app/page.tsx"
    feat = "apps/web/src/features/bookings/ui.tsx"
    envp = "/workspace/apps/web/src/env.ts"
    readme = "/workspace/README.md"

    check(
        "no-state app allow",
        gate_from_payload(_payload(app), {})["permission"],
        "allow",
    )
    check(
        "pending shape app deny",
        gate_from_payload(_payload(app), pending)["permission"],
        "deny",
    )
    check(
        "pending shape feature deny",
        gate_from_payload(_payload(feat), pending)["permission"],
        "deny",
    )
    check(
        "pending shape env allow",
        gate_from_payload(_payload(envp), pending)["permission"],
        "allow",
    )
    check(
        "pending shape readme allow",
        gate_from_payload(_payload(readme), pending)["permission"],
        "allow",
    )
    check(
        "shape done app allow",
        gate_from_payload(_payload(app), done)["permission"],
        "allow",
    )
    check(
        "ui_writes deny wins",
        gate_from_payload(_payload(app), deny_override)["permission"],
        "deny",
    )
    check(
        "ui_writes allow wins",
        gate_from_payload(_payload(app), allow_override)["permission"],
        "allow",
    )
    check(
        "unreadable yaml fail-closed",
        gate_from_payload(_payload(app), None)["permission"],
        "deny",
    )
    empty = status_digest({})
    if "no product being shaped" not in empty.lower():
        failures.append(f"empty digest: {empty!r}")
    digest = status_digest(
        load_yaml_text(
            """
phase: 2
phases:
  0: { name: salvage, status: done }
  2: { name: field, status: blocked }
  3: { name: shape, status: pending }
held:
  - id: H3
    what: Field visit to one temple
    raised: 2026-08-20
    status: open
"""
        )
    )
    for needle in ("Waiting on you", "Field visit", "Do not edit", "Harness"):
        if needle not in digest:
            failures.append(f"digest missing {needle!r}: {digest!r}")

    if failures:
        sys.stderr.write("selftest failed:\n")
        for line in failures:
            sys.stderr.write(f"  {line}\n")
        return 1
    sys.stderr.write(f"selftest passed ({len(PLAYBOOK_SKILLS)} playbook skills recorded)\n")
    return 0


def main() -> int:
    cmd = sys.argv[1] if len(sys.argv) > 1 else ""
    if cmd == "gate":
        return cmd_gate()
    if cmd == "session":
        return cmd_session()
    if cmd == "selftest":
        return cmd_selftest()
    sys.stderr.write("usage: playbook.py gate|session|selftest\n")
    return 2


if __name__ == "__main__":
    raise SystemExit(main())

File: .cursor/hooks/playbook.ts
#!/usr/bin/env bun
/**
 * Playbook harness hooks.
 *
 * Commands (stdout is JSON for Cursor; diagnostics go to stderr):
 *
 *   gate      preToolUse — deny product UI/route/feature writes until shape is done
 *   branch    beforeShellExecution — deny pushes to main and unsafe force-pushes
 *   session   sessionStart — inject the pipeline digest as additional_context
 *   selftest  run fixtures; exit 1 on failure
 *
 * Repo root is two levels above this file. Cloud and local checkouts both work.
 *
 * This was Python once, with a hand-written YAML subset parser attached, because
 * a gate that guards the build must not depend on a library being installed.
 * Bun parses YAML natively and the repo refuses to install with anything else
 * (`only-allow bun`, `engines.bun`), so the parser is gone and the dependency is
 * one the repo already had. The gate stays fail-closed: if bun is missing the
 * hook errors, and an erroring gate denies.
 */

import { existsSync, readFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";

const ROOT = resolve(dirname(Bun.fileURLToPath(import.meta.url)), "..", "..");
const STATE_PATH = join(ROOT, "docs", "product", "state.yaml");
const GATED_FRAGMENTS = ["/src/app/", "/src/features/"] as const;

const PLAYBOOK_SKILLS = [
  "next",
  "status",
  "plan",
  "build",
  "artifacts",
  "journeys",
  "design-system",
  "shape",
  "salvage",
  "setup",
  "customize",
  "ontology",
  "prototype",
  "field-kit",
  "linear-sync",
] as const;

type State = Record<string, unknown>;
type Phase = Record<string, unknown> & { _key?: string };

function emit(payload: unknown): void {
  process.stdout.write(`${JSON.stringify(payload)}\n`);
}

async function readStdin(): Promise<string> {
  try {
    return await Bun.stdin.text();
  } catch {
    return "";
  }
}

/** Parsed state, {} when there is no file, null when it cannot be read. */
export function loadState(text?: string): State | null {
  let raw = text;
  if (raw === undefined) {
    if (!existsSync(STATE_PATH)) return {};
    try {
      raw = readFileSync(STATE_PATH, "utf8");
    } catch {
      return null;
    }
  }
  try {
    const data = Bun.YAML.parse(raw) ?? {};
    if (typeof data !== "object" || Array.isArray(data)) return null;
    return data as State;
  } catch {
    return null;
  }
}

function phaseEntries(state: State): Phase[] {
  const raw = state.phases;
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) return [];
  return Object.entries(raw as Record<string, unknown>)
    .filter(([, v]) => v && typeof v === "object" && !Array.isArray(v))
    .map(([key, v]) => ({ ...(v as Record<string, unknown>), _key: key }));
}

function namedPhase(state: State, name: string): Phase | undefined {
  return phaseEntries(state).find((e) => String(e.name ?? "") === name);
}

function shapeDone(state: State): boolean {
  return String(namedPhase(state, "shape")?.status ?? "") === "done";
}

function uiWritesOverride(state: State): "allow" | "deny" | null {
  const raw = state.ui_writes;
  if (raw === undefined || raw === null) return null;
  const value = String(raw).trim().toLowerCase();
  return value === "allow" || value === "deny" ? value : null;
}

/** Whether product UI writes are allowed. No state file means boilerplate: allowed. */
export function uiWritesAllowed(state: State | null): [boolean, string] {
  if (state === null) {
    return [false, "docs/product/state.yaml is unreadable; refusing product UI writes"];
  }
  if (Object.keys(state).length === 0) {
    return [true, "no product pipeline; boilerplate edits allowed"];
  }
  const override = uiWritesOverride(state);
  if (override === "allow") return [true, "ui_writes: allow"];
  if (override === "deny") return [false, "ui_writes: deny"];
  if (shapeDone(state)) return [true, "shape is done; design doc is approved"];
  return [false, "design doc is not approved (shape is not done)"];
}

function collectPaths(node: unknown, out: string[] = []): string[] {
  if (!node || typeof node !== "object") return out;
  if (Array.isArray(node)) {
    for (const item of node) collectPaths(item, out);
    return out;
  }
  for (const [key, value] of Object.entries(node)) {
    if (
      typeof value === "string" &&
      ["path", "file_path", "filePath", "target_notebook"].includes(key)
    ) {
      out.push(value);
    } else {
      collectPaths(value, out);
    }
  }
  return out;
}

function isGatedPath(path: string): boolean {
  let normalized = path.replaceAll("\\", "/");
  if (!normalized.startsWith("/")) normalized = `/${normalized}`;
  return GATED_FRAGMENTS.some((fragment) => normalized.includes(fragment));
}

export function gateFromPayload(payload: Record<string, unknown>, state: State | null) {
  const toolInput = payload.tool_input ?? payload.input ?? {};
  const gated = collectPaths(toolInput).filter(isGatedPath);
  if (gated.length === 0) return { permission: "allow" };

  const [allowed, reason] = uiWritesAllowed(state);
  if (allowed) return { permission: "allow" };

  return {
    permission: "deny",
    user_message: `Product UI/routes/features are blocked until the design doc is approved. ${reason}.`,
    agent_message:
      "HOOK-DENY: writes under apps/*/src/app or apps/*/src/features are blocked " +
      `until shape is done (or ui_writes: allow). ${reason}. ` +
      "Do not retry the same edit or bypass it with the shell. " +
      "Work in docs/product, docs/plans, docs/journeys, or docs/research instead.",
  };
}

// Commands an agent must not run. The human can still run any of them by hand —
// that is the point. A release is a person's decision, not a slice's side effect.
const BRANCH_RULES: [RegExp, string][] = [
  [
    /\bgit\s+push\b(?=.*\bmain\b)/,
    "PRs target staging; main is production. Push to staging, then a human merges " +
      "staging into main. A hotfix onto main is theirs to run, not yours.",
  ],
  [
    /\bgh\s+pr\s+create\b(?=.*--base[= ]+main\b)/,
    "Open the PR against staging. main only ever receives a staging merge.",
  ],
  [
    /\bgit\s+push\b(?=.*(?:--force|-f)\b)(?!.*--force-with-lease)/,
    "Use --force-with-lease, and only on your own slice branch.",
  ],
];

export function branchFromCommand(command: string) {
  for (const [pattern, why] of BRANCH_RULES) {
    if (pattern.test(command)) {
      return {
        permission: "deny",
        user_message: `Blocked: ${command}`,
        agent_message:
          `${why} Do not work around this with another command — ` +
          "tell the user what you wanted to run and why.",
      };
    }
  }
  return { permission: "allow" };
}

const HARNESS_NOTE =
  "Harness: if next/status/shape are missing from your skill list, read " +
  ".agents/skills/<name>/SKILL.md (also .cursor/skills/) and say the catalog omitted them.";

async function cmdSession(): Promise<number> {
  await readStdin();
  let digest = "";
  try {
    const result = Bun.spawnSync(["bun", "scripts/status.ts"], { cwd: ROOT });
    digest = result.stdout.toString().trim();
  } catch {
    digest = "";
  }
  if (!digest) digest = "Could not render the pipeline digest — run `bun scripts/status.ts`.";
  emit({ additional_context: `${digest}\n\n${HARNESS_NOTE}` });
  return 0;
}

async function cmdGate(): Promise<number> {
  const raw = await readStdin();
  let payload: Record<string, unknown> = {};
  try {
    const parsed = raw.trim() ? JSON.parse(raw) : {};
    if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) payload = parsed;
  } catch {
    payload = {};
  }
  emit(gateFromPayload(payload, loadState()));
  return 0;
}

async function cmdBranch(): Promise<number> {
  const raw = await readStdin();
  let command = "";
  try {
    const parsed = raw.trim() ? JSON.parse(raw) : {};
    if (parsed && typeof parsed === "object") command = String(parsed.command ?? "");
  } catch {
    command = "";
  }
  emit(branchFromCommand(command));
  return 0;
}

function cmdSelftest(): number {
  const failures: string[] = [];
  const check = (name: string, got: unknown, expected: unknown) => {
    if (got !== expected)
      failures.push(`${name}: got ${JSON.stringify(got)} expected ${JSON.stringify(expected)}`);
  };

  const pending = loadState("phase: 3\nphases:\n  3: { name: shape, status: pending }\n");
  const done = loadState(
    "phase: 4\nphases:\n  3: { name: shape, status: done }\n  4: { name: journeys, status: in-progress }\n",
  );
  const denyOverride = loadState("ui_writes: deny\nphases:\n  3: { name: shape, status: done }\n");
  const allowOverride = loadState(
    "ui_writes: allow\nphases:\n  3: { name: shape, status: pending }\n",
  );

  const payload = (path: string) => ({ tool_name: "StrReplace", tool_input: { path } });
  const app = "/workspace/apps/web/src/app/page.tsx";
  const feat = "apps/web/src/features/bookings/ui.tsx";
  const envp = "/workspace/apps/web/src/env.ts";
  const readme = "/workspace/README.md";

  check("no-state app allow", gateFromPayload(payload(app), {}).permission, "allow");
  check("pending shape app deny", gateFromPayload(payload(app), pending).permission, "deny");
  check("pending shape feature deny", gateFromPayload(payload(feat), pending).permission, "deny");
  check("pending shape env allow", gateFromPayload(payload(envp), pending).permission, "allow");
  check(
    "pending shape readme allow",
    gateFromPayload(payload(readme), pending).permission,
    "allow",
  );
  check("shape done app allow", gateFromPayload(payload(app), done).permission, "allow");
  check("ui_writes deny wins", gateFromPayload(payload(app), denyOverride).permission, "deny");
  check("ui_writes allow wins", gateFromPayload(payload(app), allowOverride).permission, "allow");
  check("unreadable yaml fail-closed", gateFromPayload(payload(app), null).permission, "deny");

  const branch = (command: string) => branchFromCommand(command).permission;
  check("push to main denied", branch("git push origin main"), "deny");
  check("push HEAD:main denied", branch("git push origin HEAD:main"), "deny");
  check("push staging allowed", branch("git push origin HEAD:staging"), "allow");
  check("slice branch allowed", branch("git push -u origin agent/f2-1-reconciliation"), "allow");
  check("pr base main denied", branch("gh pr create --base main --title x"), "deny");
  check("pr base staging allowed", branch("gh pr create --base staging --title x"), "allow");
  check("bare force denied", branch("git push --force origin my-branch"), "deny");
  check(
    "force-with-lease allowed",
    branch("git push --force-with-lease origin my-branch"),
    "allow",
  );
  check("unrelated command allowed", branch("bun test"), "allow");
  check("empty command allowed", branch(""), "allow");

  // The YAML is real now, so malformed input must still fail closed rather than throw.
  check("malformed yaml is null", loadState("phases:\n  - [unclosed\n"), null);
  check("scalar yaml is null", loadState("just a string\n"), null);

  if (failures.length > 0) {
    process.stderr.write("selftest failed:\n");
    for (const line of failures) process.stderr.write(`  ${line}\n`);
    return 1;
  }
  process.stderr.write(`selftest passed (${PLAYBOOK_SKILLS.length} playbook skills recorded)\n`);
  return 0;
}

const cmd = process.argv[2] ?? "";
if (cmd === "gate") process.exit(await cmdGate());
else if (cmd === "branch") process.exit(await cmdBranch());
else if (cmd === "session") process.exit(await cmdSession());
else if (cmd === "selftest") process.exit(cmdSelftest());
else {
  process.stderr.write("usage: playbook.ts gate|branch|session|selftest\n");
  process.exit(2);
}
